accepted/2021/net6.0-tfms/net6.0-tfms.md

```diff
@@ -0,0 +1,347 @@
+* Mac Catalyst is able to consume libraries built for Xamarin iOS, even though
+  the binary might not work. The developer should get a warning, but the tools
+  shouldn't block it.
+
+### Non-Goals
+
+* Re-design the TFM syntax
+
+## Stakeholders and Reviewers
+
+* MSBuild / SDK team
+* NuGet team
+* Project system team
+* Xamarin team
+
+## Design
+
+| TFM                | Compatible With                          |
+|--------------------|------------------------------------------|
+| net6.0             | (subsequent version of net5.0)           |
+| net6.0-windows     | (subsequent version of net5.0-windows)   |
+| net6.0-android     | xamarin.android                          |
+|                    | (+everything else inherited from net6.0) |
+| net6.0-ios         | xamarin.ios                              |
+|                    | (+everything else inherited from net6.0) |
+| net6.0-macos       | xamarin.mac                              |
+|                    | (+everything else inherited from net6.0) |
+| net6.0-maccatalyst | xamarin.ios                              |
+|                    | (+everything else inherited from net6.0) |
+| net6.0-tvos        | xamarin.tvos                             |
+|                    | (+everything else inherited from net6.0) |
+
+### Compatibility rules
+
+The [built-in compatibility rules][net5.0] of the `netX.Y-platform` TFMs make a
+TFM automatically compatible with:
+
+1. Itself
+2. Earlier versions of itself
+3. The corresponding `netX.Y` TFM
+4. Earlier versions of the corresponding `netX.Y` TFM
+
+The `net5.0` TFM has the additional twist that it's compatible with
+`netcoreapp3.1` (and earlier), `netstandard2.1` (and earlier) as well as .NET
+Framework (`net4x` and earlier, but under a warning). The .NET Framework
+compatibility is achieved as a *fallback*, that is, it's only used when nothing
+else applies.
+
+We don't want to model the Xamarin TFMs as a fallback but instead want them to
+apply **before** `netcoreapp` and `netstandard`. Usually the existence of the
+Xamarin TFM implementation is to bait a `netstandard` API and switch the
+platform specific implementation. Most of the Plugin type packages use this
+pattern and in fact the `netstandard` "implementation" is more like a reference
+assembly with just stubs that throw at runtime. `Xamarin.Essentials` is a good
+example of this. The same rational can be applied to `netcoreapp` where this
+would logically be the server implementation (or maybe a .NET Core 3.x desktop
+implementation).
+
+A `net6.0-maccatalyst` project referencing a NuGet package should behave as
+follows:
+
+* It should prefer `xamarin.ios` assets over `netcoreapp` and `netstandard`
+  assets.
+    - However, it should still prefer `net5.0`/`net6.0` assets over
+      `xamarin.ios`.
+    - It also shouldn't be compatible with any other existing Xamarin TFM (such
+      as `xamarin.mac`)
+* Generate NuGet warning [NU1701] when a `xamarin.ios` asset is being used
```

Comment:
Agreed. As I said on email, I'm not particularly fond of this either. But I can get behind this because this doesn't introduce new crazy rules between the new TFMs but between a new TFM and an old TFM. I can justify that with our obligation to existing customers.

Reply:
If xamarin.ios is preferred over netcoreapp, I think that the warning is weird as well. Same reasoning as above applies, it's not really a fallback, it's just how compatibility is in this case.

Reply:
The problem I see is this:

1. Silently using the .NET Core assets is unlikely to result in a workable application due to bait & switch
2. Slenetly using xamarin.ios, while having a higher chance of working, might not work either because Mac Catalyst isn't 100% compatible with iOS.

So it seems to me the precedence rules + warning is best user experience we can deliver.

Reply:
Given 1, why not warn in the case of .NET Core assets older than 3.1 as well? 

The thing that stands out to me is the following (actual names abstracted)

Given A: 

A - works
A' - works
B - works (with a warning)
C - not likely to work (no warning). 

I know it's an oversimplification, but just curious :)

Reply:
It's not that netcoreapp/netstandard assets wouldn't work in Mac Catalyst; it's that if a package offers both xamarin.ios and netcoreapp/netstandard then the likelyhood of netcoreapp/netstandard being the right ones is low. IOW, it's probabilty isn't just a function of the project TFM and the a given offered TFM, it's about which TFM to pick given a set of TFMs. And that's what the precedence rules try to accomodate.

Does that make more sense?